projects/gloo/pkg/bootstrap/clients/secret.go

```
@@ -0,0 +1,295 @@
+			switch src := deprecatedApiSource.(type) {
+			case *v1.Settings_DirectorySecretSource:
+				{
+					newApiSource.Source = &v1.Settings_SecretOptions_Source_Directory{
+						Directory: src.DirectorySecretSource,
+					}
+				}
+			case *v1.Settings_VaultSecretSource:
+				{
+					newApiSource.Source = &v1.Settings_SecretOptions_Source_Vault{
+						Vault: src.VaultSecretSource,
+					}
+				}
+			case *v1.Settings_KubernetesSecretSource:
+				{
+					newApiSource.Source = &v1.Settings_SecretOptions_Source_Kubernetes{
+						Kubernetes: src.KubernetesSecretSource,
+					}
+				}
+			}
+			secretOpts.SecretSources = []*v1.Settings_SecretOptions_Source{newApiSource}
+		}
+	}
+	if secretOpts != nil {
+		return NewMultiResourceClientFactory(secretOpts.GetSecretSources(), sharedCache, cfg, clientset, kubeCoreCache, vaultClient)
+	}
+	return nil, errors.Errorf("invalid config source type")
+}
+
+type MultiResourceClientFactory struct {
+	secretSources []*v1.Settings_SecretOptions_Source
+	sharedCache   memory.InMemoryResourceCache
+	cfg           **rest.Config
+	clientset     *kubernetes.Interface
+	kubeCoreCache *cache.KubeCoreCache
+	vaultClient   *vaultapi.Client
+
+	refreshRate     *durationpb.Duration
+	watchNamespaces []string
+}
+
+var (
+	// ErrNilSourceSlice indicates a nil slice of sources was passed to the factory,
+	// and we can therefore not initialize any sub-clients
+	ErrNilSourceSlice = errors.New("nil slice of secretSources")
+)
+
+// NewMultiResourceClientFactory returns a resource client factory for a client
+// supporting multiple sources
+func NewMultiResourceClientFactory(secretSources []*v1.Settings_SecretOptions_Source,
+	sharedCache memory.InMemoryResourceCache,
+	cfg **rest.Config,
+	clientset *kubernetes.Interface,
+	kubeCoreCache *cache.KubeCoreCache,
+	vaultClient *vaultapi.Client) (*MultiResourceClientFactory, error) {
+
+	// Guard against nil source slice
+	if secretSources == nil {
+		return nil, ErrNilSourceSlice
```

I need to find out what the behavior of returning this error is and if we should attempt to use a default secret source. I think if we crashloop the pod on this error, that's pretty bad. I don't know what default we could safely presume to use though or if that would hide the issue from the user